### PR TITLE
Only re-deploy when code or service updated

### LINKED ISSUE #2710: DX: Only re-deploy Lambda function if code actually changes
# This is a Feature Proposal

## Description
We could limit deployment / creation of new lambda functions, by checking if the lambda function did actually change.
Proposal:
- variant one: hash the contents of the zipfile / use hash as file name and check if file already exists on s3.
- variant two (prerequisite: shrinkwrap.json or yarn.lock): check if all dependencies are locked (see prerequisites) and then run a hash on the lockfile + source code to be used as file name and only upload the zip file if the hash does not already exist on s3.
- variant three: just run a hash on package.json and source code as zip file name (doesn't have the guarantees of proposal nr.1 (predictable transient dependencies), but might still be a viable solution).
- variant four: leverage existing vcs to get the last hash of a specific service folder / file to verify that it did not change.
I think variant one would be most viable and work for every language.

Similar or dependent issues:
* #2325

### PR DESCRIPTION
**Pull request by pmuens**
## What did you implement:

Closes #2710

Checks if the function code or service configuration has changed and only deploys if that's the case.

## How did you implement it:

Add a `sha256` based hash comparison for the files in question (e.g. the `.zip` or `CloudFormation` files). Only deploy when the hashes are different.

This PR introduces a new AWS related lifecycle event (`checkForChanges`) and sets a globally available flag (`this.serverless.service.provider.shouldNotDeploy`) so that others can use the result of the check as well. Thanks for pointing that out @HyperBrain 👍 

## How can we verify it:

Use the following service:

```yml
service:
  name: my-service

provider:
  name: aws
  runtime: nodejs6.10

functions:
  hello:
    handler: handler.hello
  goodbye:
    handler: handler.goodbye
```

1. Run `serverless deploy` to deploy everything.
2. Run `serverless deploy function -f hello` --> Console will tell you that nothing was deployed because the code was not changed.
3. Update `hello` function code
4. Run `serverless deploy function -f hello` --> New function code will be uploaded
5. Run `serverless deploy`
6. Run `serverless deploy` (yes, a second time) --> Nothing will be updated since nothing changed
7. Change the `hello` function code
8. Run `serverless deploy` --> Deployment should be triggerend since the function code changed
9. Run `serverless deploy` --> Nothing should change since nothing was updated
10. Change configuration by e.g. adding a new function
11. Run `serverless deploy` --> Deployment should succeed

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

**Comment by pmuens** [comment]
@eahefnawy this is ready from an implementation perspective (I only need to add some tests).

Could you please review and test it thoroughly the upcoming days?

**Comment by HyperBrain** [comment]
@pmuens @eahefnawy I'm not sure if that will work. If I only change settings for some resources that exist in my serverless.yml (like the provisioning for DynamoDB tables), it will not deploy until I also change some code?
This would not be correct. Any change in the CF template's resources should lead to a redeploy, not only code changes. For "deploy function" it is debatable if the behavior is correct or not, but for "deploy" it must deploy in any case.

**Comment by pmuens** [comment]
> @pmuens @eahefnawy I'm not sure if that will work. If I only change settings for some resources that exist in my serverless.yml (like the provisioning for DynamoDB tables), it will not deploy until I also change some code?
> This would not be correct. Any change in the CF template's resources should lead to a redeploy, not only code changes. For "deploy function" it is debatable if the behavior is correct or not, but for "deploy" it must deploy in any case.

Thanks for the feedback @HyperBrain 👍 

The title / description might be misleading here. Here's how it's implemented:

**`deploy function`**

- Compute local hash and compare against `CodeSha256` of function
- Re-Deploy if hash is different

**`deploy`**

- Compute hashes for `.zip` files
- Compute hash for in memory representation of CloudFormation template
- Fetch remote hashes (they're now added as metadata to the S3 objects)
- Do a `_.difference` of the remote hashes against the local hashes
- Re-deploy if a difference was found (regardless if it's a zip file or the CloudFormation template)

This way it will re-deploy if the zip file(s) change(s) or the CloudFormation template changes. Both will trigger a re-deploy. The deployment is only skipped if the `.zip` file(s) are untouched and the CloudFormation file is untouched.

**Comment by HyperBrain** [comment]
@pmuens Thanks for the explanation, now it's clear 😃 

But, could we put the verification into a separate lifecycle `checkForChanges`, just before `upload....`, so that plugins can hook in there and overwrite the check or add their own checks? E.g. the alias plugin should add it's own validation and allow a redeploy in case also the alias stack changed (and not only the base stack).

**Comment by pmuens** [comment]
@HyperBrain thanks for confirming 👍 

> But, could we put the verification into a separate lifecycle checkForChanges, just before upload...., so that plugins can hook in there and overwrite the check or add their own checks? E.g. the alias plugin should add it's own validation and allow a redeploy in case also the alias stack changed (and not only the base stack).

That sounds like a good idea. Would we still be able to skip the deployment if we do that? Right now I've added a pretty ugly check in the `deploy:deploy` lifecycle which resolves if nothing has changed (and therefore skips the whole deployment phase). How would this look like / work if we add a new lifecycle event?

**Comment by HyperBrain** [comment]
In the separate `checkForChanges` event you could check for a `force` flag. Plugins that need the check to be skipped could then set the flag to disable SLS's internal check and force the deploy phase to be triggered regardless of differences or not. The flag can be named so that it is clear that it is for internal use, or it even could be made public, so that a `deploy --force` will skip the check on the user's demand.

File: lib/plugins/aws/deploy/index.js [hunk 1]
```diff
@@ -32,8 +33,9 @@ class AwsDeploy {
     Object.assign(
       this,
       extendedValidate,
-      createStack,
       setBucketName,
+      shouldBeDeployed,
```
Comment (HyperBrain) [inline lib/plugins/aws/deploy/index.js]: Somehow I disagree with the naming - all other methods are named after the actions they do. `shouldBeDeployed` doesn't really fit in there. Maybe `checkForDifferences` would be more appropriate.

File: lib/plugins/aws/deploy/index.js [hunk 2]
```diff
@@ -85,7 +87,16 @@ class AwsDeploy {
           if (this.options.noDeploy) {
             return BbPromise.resolve();
           }
-          return this.serverless.pluginManager.spawn('aws:deploy:deploy');
+
+          return BbPromise.bind(this)
+            .then(this.setBucketName)
+            .then(this.shouldBeDeployed)
```
Comment (HyperBrain) [inline lib/plugins/aws/deploy/index.js]: I would add a new event `checkForChanges` in the commands:lifecycleEvents just before `uploadArtifacts` and do the check for differences in `aws:deploy:deploy:checkForChanges`.
If the check is not disabled (by something like a `--force` switch) the event should bail out with a rejection, that tells the user that nothing has been deployed.

This will keep the outer event lifecycle simple and understandable. Imo the inner lifecycle should be started in any case, but should bail out if it is terminated due to the condition.

File: lib/plugins/aws/deploy/index.js (hunk 1 above)
Comment (pmuens) [inline lib/plugins/aws/deploy/index.js]: Good one. TBH I'm also slightly embarrassed with the name.

I was literally sitting there yesterday trying to come up with an appropriate name.

But you know:

> There are only two hard things in Computer Science, ...

File: lib/plugins/aws/deploy/index.js (hunk 2 above)
Comment (HyperBrain) [inline lib/plugins/aws/deploy/index.js]: Instead of rejecting, the checkForChanges method/event could set a property in the deploy plugin (like this.skipPhysicalDeployments) that will be evaluated by the `uploadArtifacts` and `updateStack` events, so that they only resolve instead of doing any physical uploads or updates.

File: lib/plugins/aws/deploy/index.js (hunk 1 above)
Comment (eahefnawy) [inline lib/plugins/aws/deploy/index.js]: lol never bored of this conversation 😄

**Comment by eahefnawy** [comment]
@pmuens Thanks a lot for that! 😊 ... I'm getting started with deep testing, but I keep getting this error when deploying a new service for the first time using this PR:

```
Eslams-MacBook-Pro:general eslam$ sldev deploy
Serverless: Packaging service...
 
  Serverless Error ---------------------------------------
 
  Stack 'sls-no-deploy-2-dev' does not exist

```

Can you reproduce? have you tried deploying with a fresh service?

Some edge cases I wanna test:

- What if config file changes, but code & CF file haven't? should we deploy?
- What if we're packaging individually?
- What if we're using a custom deployment package?

File: lib/plugins/aws/deployFunction/index.js [hunk 3]
```diff
@@ -15,12 +16,14 @@ class AwsDeployFunction {
       path.join(this.serverless.config.servicePath || '.', '.serverless');
     this.provider = this.serverless.getProvider('aws');
 
+    // used to store data received via AWS SDK
+    this.remoteFunctionData = null;
```
Comment (pmuens) [inline lib/plugins/aws/deployFunction/index.js]: This should be updated to `this.serverless.service.provider.remoteFunctionData`

File: lib/plugins/package/lib/zipService.js [hunk 4]
```diff
@@ -90,6 +90,7 @@ module.exports = {
           zip.append(fs.readFileSync(fullPath), {
             name: filePath,
             mode: stats.mode,
+            date: new Date(0), // necessary to get the same hash when zipping the same content
```
Comment (pmuens) [inline lib/plugins/package/lib/zipService.js]: We should add a test for that to ensure that this specific date is always set.

**Comment by HyperBrain** [comment]
I will update the [lifecycle cheat sheet](https://gist.github.com/HyperBrain/50d38027a8f57778d5b0f135d80ea406) to include the new `checkForChanges` event as soon as this has been merged.

**Comment by pmuens** [comment]
> I will update the lifecycle cheat sheet to include the new checkForChanges event as soon as this has been merged.

Thanks @HyperBrain 🙌

**Comment by pmuens** [comment]
> The only edge case that I found is when you're referencing env vars in the custom section of the config file. This causes the deployment to be skipped even if the value of the env vars has changed, which yields a change in the config file.

@eahefnawy That's really a good find! 🤔 

Maybe we can add a `--force` option or smth. to force a deployment even if nothing has changed?

Any thoughts?

**Comment by pmuens** [comment]
**Quick update:** Just added support for the `--force` option. This way one can ensure that the deployment will be done (even if the files haven't changed).

**Comment by pmuens** [comment]
Another update:

The bug which causes `package.individually` to skip the deployment has been fixed.

However the problem with the `rollback` is still present. IMO this is a bug in rollback since it should set the rolled back deployment to the most recent one. So we should fix it in `rollback` and not here.

**Comment by eahefnawy** [comment]
@pmuens agreed with your remarks. I think this PR is G2M from my side. Is it G2M from your side @pmuens ?

**Comment by pmuens** [comment]
> @pmuens agreed with your remarks. I think this PR is G2M from my side. Is it G2M from your side @pmuens ?

@eahefnawy yes, I tested it thoroughly today with different setups and services (running this code on old deployment w/o metadata, activating `package.individually` etc.). :shipit:
